File: pathfinder-core/src/main/java/de/cubbossa/pathfinder/node/implementation/PlayerNode.java
package de.cubbossa.pathfinder.node.implementation;

import de.cubbossa.pathfinder.AbstractPathFinder;
import de.cubbossa.pathfinder.Changes;
import de.cubbossa.pathfinder.misc.Location;
import de.cubbossa.pathfinder.misc.PathPlayer;
import de.cubbossa.pathfinder.node.AbstractNodeType;
import de.cubbossa.pathfinder.node.Edge;
import de.cubbossa.pathfinder.node.Node;
import de.cubbossa.pathfinder.node.NodeType;
import java.util.Collection;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Getter
public class PlayerNode implements Node {

  public static final NodeType<PlayerNode> TYPE = new AbstractNodeType<>(
      AbstractPathFinder.pathfinder("player")
  ) {

    @Override
    public boolean canBeCreated(Context context) {
      return false;
    }

    @Override
    public PlayerNode createNodeInstance(Context context) {
      throw new IllegalStateException("PlayerNodes are only part of runtime navigation and "
          + "must be created from constructor.");
    }

    @Override
    public @Nullable PlayerNode createAndLoadNode(Context context) {
      throw new IllegalStateException("PlayerNodes are only part of runtime navigation and "
          + "must be created from constructor.");
    }
  };

  private final PathPlayer<?> player;
  private final Changes<Edge> edgeChanges = new Changes<>();

  public PlayerNode(PathPlayer<?> player) {
    this.player = player;
  }

  @Override
  public UUID getNodeId() {
    return player.getUniqueId();
  }

  @Override
  public Location getLocation() {
    return player.getLocation();
  }

  @Override
  public void setLocation(Location location) {

  }

  @Override
  public Collection<Edge> getEdges() {
    return new HashSet<>();
  }

  @Override
  public Optional<Edge> connect(UUID other, double weight) {
    return Optional.empty();
  }

  @Override
  public int compareTo(@NotNull Node o) {
    return 0;
  }

  @Override
  public Node clone() {
    try {
      return (Node) super.clone();
    } catch (CloneNotSupportedException e) {
      return new PlayerNode(player);
    }
  }

  @Override
  public Node clone(UUID id) {
    throw new IllegalStateException("Cannot clone a player node with ID parameter.");
  }

  @Override
  public String toString() {
    return "PlayerNode{" +
        "player=" + player.getName() +
        '}';
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) {
      if (o instanceof Node node) {
        return node.getNodeId().equals(getNodeId());
      }
      return false;
    }
    PlayerNode that = (PlayerNode) o;
    return Objects.equals(getPlayer().getUniqueId(), that.getPlayer().getUniqueId());
  }

  @Override
  public int hashCode() {
    return getPlayer().getUniqueId().hashCode();
  }
}

File: pathfinder-core/src/main/java/de/cubbossa/pathfinder/node/implementation/PlayerNode.kt
package de.cubbossa.pathfinder.node.implementation

import de.cubbossa.pathfinder.AbstractPathFinder
import de.cubbossa.pathfinder.Changes
import de.cubbossa.pathfinder.misc.Location
import de.cubbossa.pathfinder.misc.PathPlayer
import de.cubbossa.pathfinder.node.AbstractNodeType
import de.cubbossa.pathfinder.node.Edge
import de.cubbossa.pathfinder.node.Node
import de.cubbossa.pathfinder.node.NodeType
import lombok.Getter
import java.util.*

@Getter
class PlayerNode(private val player: PathPlayer<*>) : Node, Cloneable {
    override val edgeChanges: Changes<Edge> = Changes()

    override val nodeId: UUID
        get() = player.uniqueId

    override var location: Location
        get() = player.location
        set(location) {
        }

    override val edges: MutableCollection<Edge>
        get() = HashSet()

    override fun connect(other: UUID, weight: Double): Edge? {
        return null
    }

    override fun compareTo(o: Node): Int {
        return 0
    }

    override fun clone(): Node {
        return PlayerNode(player)
    }

    override fun clone(id: UUID): Node {
        throw IllegalStateException("Cannot clone a player node with ID parameter.")
    }

    override fun toString(): String {
        return "PlayerNode{" +
                "player=" + player.name +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) {
            if (other is Node) {
                return other.nodeId == nodeId
            }
            return false
        }
        val that = other as PlayerNode
        return player.uniqueId == that.player.uniqueId
    }

    override fun hashCode(): Int {
        return player.uniqueId.hashCode()
    }

    companion object {
        val TYPE: NodeType<PlayerNode> = object : AbstractNodeType<PlayerNode>(
            AbstractPathFinder.pathfinder("player")
        ) {
            override fun canBeCreated(context: NodeType.Context): Boolean {
                return false
            }

            override fun createNodeInstance(context: NodeType.Context): PlayerNode {
                throw IllegalStateException(
                    "PlayerNodes are only part of runtime navigation and "
                            + "must be created from constructor."
                )
            }

            override fun createAndLoadNode(context: NodeType.Context): PlayerNode? {
                throw IllegalStateException(
                    "PlayerNodes are only part of runtime navigation and "
                            + "must be created from constructor."
                )
            }
        }
    }
}
